### DISCUSSION TITLE
How to migrate from using SourceAsset to AssetSpec

### DISCUSSION THREAD

**Discussion by ShootingStarD** [dagster-io/dagster#24005] [Q&A] [answered]
### What's the issue or suggestion?

Hi, in the release of dagster 1.8 it is said that SourceAsset are being deprecated and we should use AssetSpec instead. My question is : how to migrate from one to the other?
Currently I face the problem that I have hyperparameter for an ML model that should depend on whether I am in a test environment, an environment with GPU or without GPU. I thought of having a json as an Asset that could be different for each environment. With a source asset it is very easy, I define my json io manager and I can easily load this json file that can be different for each environment
However I get the feeling that with AssetSpec, we cannot really load the "Asset" into another asset (as we cannot define an IO manager)
Therefore how should we migrate from SourceAsset to AssetSpec if AssetSpec have less functionalities than SourceAsset?

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by ShootingStarD** [comment]
I have another example where I don't really see how to solve it. We have a machine learning model that must be read in memory using our xgboost io manager. The model is made manually in jupyter notebooks because we could change the hyperparameter, the features etc. With a source Asset , we can specify the io manager to use to load the model, but I do not see how to load the model in another asset using the AssetSpec

**Comment by HynekBlaha** [comment, marked as answer]
Hello @ShootingStarD, have you tried steps mentioned in [migration docs](https://docs.dagster.io/migration)?
```python
# before
from dagster import SourceAsset
my_asset = SourceAsset("my_asset", io_manager_key="abc")

# after
from dagster import AssetSpec
my_asset = AssetSpec("my_asset", metadata={"dagster/io_manager_key": "abc"})
```

**Reply by ShootingStarD** [reply]
Ok so I tested it and it can work however I realize than the [load_assets_from_package_module()](https://docs.dagster.io/_apidocs/assets#dagster.load_assets_from_package_module) does not return AssetScec, like it does for the SourceAsset, so I have to extend the returned list with all my assets specs. Is it the intended behaviour? Could we make it so that it also loads AssetSpec?

**Reply by zaneselvans** [reply]
I am running into the same problem when trying to migrate to `AssetSpec` -- how can one easily collect a bunch of `AssetSpec` from a module when building a Definition?

**Reply by deepyaman** [reply]
@ShootingStarD @zaneselvans I just ran into this problem, too.

> Is it the intended behaviour?

I'm guessing not. :)

> Could we make it so that it also loads AssetSpec?

~Drafting a PR now!~ **Update**: Sorry, it looks like I was over-eager in taking this; the change is more complex than I understood, but somebody else is looking into it!

**Reply by lorenzotenti** [reply]
Hello! Is there any news on this? I'm also migrating to 1.8 and moving from `SourceAsset` to `AssetSpec`, but it looks like assets defined with AssetSpec are not returned in `AssetSelection.all()`.

**Reply by ion-elgreco** [reply]
@lorenzotenti you need to add include_sources afaik

**Comment by ShootingStarD** [comment]
Oh, sorry I didn't see the migration guide, thanks I'll follow that
